geps/gep-2895/index.md

```diff
@@ -0,0 +1,265 @@
+
+* Issue: [#2895](https://github.com/kubernetes-sigs/gateway-api/issues/2895)
+* Status: Experimental
+
+## TLDR
+
+Just like modify header is useful, the same goes for query parameters. 
+It's helpful to have a `HTTPQueryParamFilter` field in `HTTPRouteFilter` to set, 
+add and remove a query parameter of the HTTP request before it is sent to the upstream target.
+
+## Goals
+
+* Provide a way to modify query parameters of an incoming request in a `HTTPRoute`.
+
+## Introduction
+
+The query parameters are an important part of the request URL. 
+The developers can use query parameters to filter, sort or customize data of request body. 
+Backend service can enable different function based on the query parameters. 
+Query parameters are important information about search and track. 
+Moreover, query parameter, headers and cookies are common techniques used in a canary release.
+
+The `HTTPRouteFilter` API now supports filters `RequestHeaderModifier` and `ResponseHeaderModifier`. 
+This GEP proposes adding support for modifying query parameters in a `HTTPRoute`.
+
+## API
+
+This GEP proposes to add a new field `HTTPQueryParamFilter` to `HTTPRouteFilter`.
+
+The `HTTPQueryParamFilter` is considered an extended feature.
+
+```golang
+const (
+    // HTTPRouteFilterQueryParamModifier can be used to set, add or remove a query
+    // parameter from an HTTP request before it is sent to the upstream target.
+    //
+    // Support in HTTPRouteRule: Extended
+    //
+    // Support in HTTPBackendRef: Extended
+    HTTPRouteFilterQueryParamModifier HTTPRouteFilterType = "QueryParamModifier"
+)
+
+// HTTPQueryParamFilter defines a filter that modifies HTTP query parameter.
+// Only one action for a given query param name is permitted.
+// Filters specifying multiple actions of the same or different type for any one
+// query param name are invalid and will be rejected by CRD validation.
+type HTTPQueryParamFilter struct {
+    // Set overwrites the HTTP request with the given query param (name, value)
+    // before the action. 
+    // The request query parameter names are case-sensitive.
+    // This must be an exact string match of query param name.
+    // (See https://www.rfc-editor.org/rfc/rfc7230#section-2.7.3).
+    //
+    // Input:
+    //   GET /foo?my-parameter=foo HTTP/1.1
+    //
+    // Config:
+    //   set:
+    //   - name: "my-parameter"
+    //     value: "bar"
+    //
+    // Output:
+    //   GET /foo?my-parameter=bar HTTP/1.1
+    //
+    // If the query parameter is not set in the request,
+    // the Set action MUST be ignored by the Gateway.
+    //
+    // Input:
+    //   GET /foo HTTP/1.1
+    //
+    // Config:
+    //   set:
+    //   - name: "my-parameter"
+    //     value: "bar"
+    //
+    // Output:
+    //   GET /foo HTTP/1.1
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=name
+    // +kubebuilder:validation:MaxItems=16
+    Set []HTTPHeader `json:"set,omitempty"`
```

Comment:
There are two differences between Header and Query string:
1. Header is case insensitive while Query string is not.
2. The value of Header can not be empty but the one in Query string can be empty.

Would be better to use a new type.

Reply:
OK. I'm checking it specifically.

Reply:
@spacewander 

Based on the Gateway v1.0.0 and v1.1.0, the type `HeaderName` is the name of a header or query parameter.

So, it's better to commit a new PR about it.
However, this GEP-2895 will add specific comments about name and value of query parameter. 

```go
// HeaderName is the name of a header or query parameter.
//
// +kubebuilder:validation:MinLength=1
// +kubebuilder:validation:MaxLength=256
// +kubebuilder:validation:Pattern=`^[A-Za-z0-9!#$%&'*+\-.^_\x60|~]+$`
// +k8s:deepcopy-gen=false
type HeaderName string
```

Reply:
I see